sdks/go/pkg/beam/pardo.go

```diff
@@ -64,9 +66,16 @@ func TryParDo(s Scope, dofn interface{}, col PCollection, opts ...Option) ([]PCo
 	}
 
 	var rc *coder.Coder
+	// Sdfs will always encode restrictions as KV<restriction, watermark state | nil>
 	if fn.IsSplittable() {
 		sdf := (*graph.SplittableDoFn)(fn)
-		rc, err = inferCoder(typex.New(sdf.RestrictionT()))
+		restT := typex.New(sdf.RestrictionT())
+		// If no watermark estimator state, use boolean as a placeholder
```

Comment:
@lostluck - I'd appreciate input on this, is there an easy way to encode nil state? Using a bool here isn't a huge deal performance-wise, but its kinda odd and not really ideal.

Reply:
Outside of specifying a new custom coder, the coder for GlobalWindows encodes them 0 size, as it's a mono value type.

I suppose this related to my question during design, about whether we're using it all the time or not. A Watermark Estimator with no state along with an SDF with type S should make the Restriction type S, rather than trying to make up a type of WS to make KV<S, WS> valid.

Reply:
Now that I'm looking at your comment in it's code context: I'm not happy with making up a type, especially if we need to re-check that decision at the execution layer. The execution layers need to make as few decisions as possible, and if they do need to be made, they should happen at Bundle construction time, not on a per call basis.

If we have to, we should make it a GlobalWindow which encodes down to 0 size, but we didn't make that easy to do at this user layer. We'd almost rather propagate the current information through to [graphx/translate.go](https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/core/runtime/graphx/translate.go#L443) and make the decisions there, since we would be able to set the coder type to a [Global window coder]( https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/core/graph/coder/windows.go#L45), but we didn't make window coders fully modular to ordinary coders, so additional work is there.

That seems hackier than trusting the restriction coder is correct by construction time (after passing through this, and later steps) and determining if we need to execute Watermark code based on the DoFn in [exec/translate.go](https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/core/runtime/exec/translate.go#L452) configuring the affected exec nodes accordingly at construction.

What do you think about that? Would that minimize the execution time branches?

Reply:
> What do you think about that? Would that minimize the execution time branches?

I'm not necessarily opposed to doing this, but I _don't_ think that it will minimized execution time branches, in fact I think it will moderately increase the number of them. Right now, we branch any time we might need to do some watermark related work - I actually don't think any of those branches would go away, and we'd need to add some additional conditional logic for reading/writing restrictions (since we wouldn't know if the restriction is located at `elm.Elm.(*FullValue).Elm2` or `elm.Elm.(*FullValue).Elm2.(*FullValue).Elm`). 

Like I said, I'm not opposed to doing this differently, but I'd lean towards adding a custom nil value coder (probably in a future PR) over sometimes encoding and sometimes not encoding watermark state. I'd actually probably lean towards the current implementation over both, though, since a single bool value doesn't add much space overhead, and introducing another custom coder actually adds a little additional overhead of its own I think. FWIW, I believe the Java Sdk just has a default watermark state timestamp that it sends when no state is actually specified (not too different from this approach)

Reply:
I updated to use a default invoker implementation that just returns false like we talked about offline - I agree that is cleaner.